esp/logging/loggingagent/cassandraloggingagent/cassandralogagent.cpp

```diff
@@ -0,0 +1,621 @@
+    {
+        setCassandraLogAgentOption(opts, "user", dbUserID.str());
+        setCassandraLogAgentOption(opts, "password", dbPassword.str());
+    }
+
+    cassSession->setOptions(opts);
+
+    //prepare defaultDB
+    ensureKeySpace();
+
+    //prepare transSeed tables
+    initTransSeedTable();
+
+    //Read logging transaction seed
+    queryTransactionSeed(loggingTransactionApp.get(), transactionSeed);
+}
+
+void CCassandraLogAgent::ensureKeySpace()
+{
+    CassandraSession s(cass_session_new());
+    CassandraFuture future1(cass_session_connect(s, cassSession->queryCluster()));
+    future1.wait("connect without keyspace");
+
+    VStringBuffer st("CREATE KEYSPACE IF NOT EXISTS %s WITH replication = { 'class': 'SimpleStrategy', 'replication_factor': '1' };",
+        defaultDB.str());
+    CassandraStatement statement(cass_statement_new(st.str(), 0));
+    CassandraFuture future2(cass_session_execute(s, statement));
+    future2.wait("execute");
+
+    s.set(NULL);
+}
+
+void CCassandraLogAgent::initTransSeedTable()
+{
+    //Create transaction seed table as needed
+    StringBuffer transSeedTableKeys;
+    StringArray transSeedTableColumnNames, transSeedTableColumnTypes;
+    transSeedTableColumnNames.append("id");
+    transSeedTableColumnTypes.append("int");
+    transSeedTableColumnNames.append("application");
+    transSeedTableColumnTypes.append("varchar");
+    transSeedTableColumnNames.append("update_time");
+    transSeedTableColumnTypes.append("timestamp");
+    transSeedTableKeys.set("application"); //primary key
+
+    setKeySpace(defaultDB.str());
+    cassSession->connect();
+    createTable(defaultDB.str(), "transactions", transSeedTableColumnNames, transSeedTableColumnTypes, transSeedTableKeys.str());
+
+    unsigned transactionCount = 0;
+    if (executeSimpleSelectStatement("SELECT COUNT(*) FROM transactions", transactionCount) == 0)
+    {
+        VStringBuffer st1("INSERT INTO transactions (id, application, update_time) values ( 10000, '%s', toUnixTimestamp(now()));", loggingTransactionApp.get());
+        VStringBuffer st2("INSERT INTO transactions (id, application, update_time) values ( 10000, '%s', toUnixTimestamp(now()));", defaultTransactionApp.get());
+        executeSimpleStatement(st1.str());
+        executeSimpleStatement(st2.str());
+    }
+    cassSession->disconnect();
+}
+
+void CCassandraLogAgent::queryTransactionSeed(const char* appName, StringBuffer& seed)
+{
+    CriticalBlock b(transactionSeedCrit);
+
+    unsigned seedInt = 0;
+    VStringBuffer st("SELECT id FROM transactions WHERE application = '%s'", appName);
+    setKeySpace(defaultDB.str()); //Switch to defaultDB since it may not be the current keyspace.
+    cassSession->connect();
+    executeSimpleSelectStatement(st.str(), seedInt);
+    seed.setf("%d", seedInt);
+
+    //update transactions for the next seed
+    VStringBuffer updateQuery("UPDATE transactions SET id=%d WHERE application = '%s'", ++seedInt, appName);
```

Comment:
I'm concerned about multiple agents updating this same "transaction id" and overriding each other's seed.